consensus/core/src/base_committer.rs

```diff
@@ -0,0 +1,371 @@
+        }
+    }
+
+    /// Check whether the specified leader has 2f+1 non-votes (blames) to be directly skipped.
+    fn enough_leader_blame(&self, voting_round: Round, leader: AuthorityIndex) -> bool {
+        let voting_blocks = self.block_store.get_blocks_by_round(voting_round);
+
+        let mut blame_stake_aggregator = StakeAggregator::<QuorumThreshold>::new();
+        for voting_block in &voting_blocks {
+            let voter = voting_block.reference().author;
+            if voting_block
+                .ancestors()
+                .iter()
+                .all(|ancestor| ancestor.author != leader)
+            {
+                tracing::trace!(
+                    "[{self}] {voting_block:?} is a blame for leader {}",
+                    AuthorityRound::new(leader, voting_round - 1)
+                );
+                if blame_stake_aggregator.add(voter, &self.committee) {
+                    return true;
+                }
+            }
+        }
+        false
+    }
+
+    /// Check whether the specified leader has 2f+1 certificates to be directly
+    /// committed.
+    fn enough_leader_support(&self, decision_round: Round, leader_block: &Block) -> bool {
+        let decision_blocks = self.block_store.get_blocks_by_round(decision_round);
+
+        // Quickly reject if there isn't enough stake to support the leader from
+        // the potential certificates.
+        let total_stake: Stake = decision_blocks
+            .iter()
+            .map(|b| self.committee.stake(b.author()))
+            .sum();
+        if total_stake < self.committee.quorum_threshold() {
+            tracing::debug!(
+                "Not enough support for: {}. Stake not enough: {} < {}",
+                leader_block.round(),
+                total_stake,
+                self.committee.quorum_threshold()
+            );
+            return false;
+        }
+
+        let mut certificate_stake_aggregator = StakeAggregator::<QuorumThreshold>::new();
+        let mut all_votes = HashMap::new();
+        for decision_block in &decision_blocks {
+            let authority = decision_block.reference().author;
+            if self.is_certificate(decision_block, leader_block, &mut all_votes) {
+                tracing::trace!(
+                    "[{self}] {decision_block:?} is a certificate for leader {leader_block:?}"
+                );
+                if certificate_stake_aggregator.add(authority, &self.committee) {
+                    return true;
+                }
+            } else {
+                tracing::trace!(
+                    "[{self}] {decision_block:?} is not a certificate for leader {leader_block:?}"
+                );
+            }
+        }
+        false
+    }
+
+    /// Apply the indirect decision rule to the specified leader to see whether
+    /// we can indirect-commit or indirect-skip it.
+    #[tracing::instrument(skip_all, fields(leader = %leader))]
+    pub fn try_indirect_decide<'a>(
```

Comment:
Let's move the `pub` methods on the top and leave the private ones on the bottom.  That will be easier for someone to read the public API of this construct.